### PR TITLE
Optim enhance

### PR DESCRIPTION
**Pull request by orenri**
* Implemented a numerical stable version for the loss and gradient of Logistic Regression, used during MPLE optimization.
Magic numbers are taken from `sklearn`:
https://github.com/scikit-learn/scikit-learn/blob/72b35a46684c0ecf4182500d3320836607d1f17c/sklearn/_loss/_loss.pyx.tp#L728
This solves the scaling problem we had (more robust to different scales of regressors).
* Transitioned to `scipy.optimize.minimize` for Logistic Regression optimization (rather than the custom optimization loop we had).
* Changed `initial_thetas` to be a dictionary (rather than a `np.ndarray`), like what is returned from `ERGM.get_model_parameters()`. Now given parameters are aligned according to the keys in the dictionary, rather than position in the array, the `collinearity_fixer` doesn't run for models that are initialized with parameters, and redundant features are removed from 'ergm._metrics_collection` on init (features that are missing in the received dictionary of initial parameters).
* Added a method `ERGM.get_mple_prediction` that serves as a getter for `ERGM._exact_average_matrix`. It calculates the MPLE prediction if the matrix doesn't exist (`ERGM._mple` didn't run). For dyadic dependent models, where the MPLE prediction is an estimation that depends on the `observed_network` provided, the prediction is calculated every time.
* Distributed MPLE prediction is not working properly, probably due to parameter parsing issues.

### CODE REVIEW

File: pyERGM/metrics.py [hunk 1]
```diff
@@ -1077,8 +1079,21 @@ def calc_statistics_for_binomial_tensor_distributed(self, tensor_size, p=0.5, nu
...
         return whole_sample_statistics
 
+    def remove_feature_by_idx(self, idx: int):
+        metric_of_feat = self.get_metric_by_feat_idx(idx)
+        is_trimmable = metric_of_feat._get_effective_feature_count() > 1
+        if not is_trimmable:
+            print(f"Removing the metric {str(metric_of_feat)} from the collection")
+            sys.stdout.flush()
+            self._delete_metric(metric=metric_of_feat)
+        else:
```
Comment (tomtalp) [inline pyERGM/metrics.py:1089]: I wonder if there's ever a use-case in which a metric is trimmable, and we delete all it's sub-indices so that we're practically left with no metric. So the metric is still there (`self._delete_metric` was never called), but the `indices_to_ignore` list is full with all indices

File: pyERGM/metrics.py [hunk 2]
```diff
@@ -1144,20 +1164,10 @@ def collinearity_fixer(self, sample_size=1000, nonzero_thr=10 ** -1, ratio_thres
...
-
-                    sample_features = np.delete(sample_features, removal_order[0], axis=0)
-
-                else:
-                    idx_to_delete = self.get_feature_idx_within_metric(removal_order[i])
-                    print(f"Removing the {idx_to_delete} feature of {str(cur_metric)} to fix multi-collinearity")
-                    sys.stdout.flush()
-                    cur_metric.update_indices_to_ignore([idx_to_delete])
+                    i = 0
+                self.remove_feature_by_idx(removal_order[i])
```
Comment (tomtalp) [inline pyERGM/metrics.py:1168]: nice

File: pyERGM/utils.py [hunk 3]
```diff
@@ -3,8 +3,10 @@
 from typing import Collection
 import numpy as np
 import networkx as nx
+from PIL.GimpGradientFile import linear
```
Comment (tomtalp) [inline pyERGM/utils.py:6]: what's this? `linear` doesn't seem to be used + `PIL`/`pillow` not in requirements.txt

File: pyERGM/metrics.py (hunk 1 above)
Comment (tomtalp) [reply pyERGM/metrics.py:1089]: This is already handled - when a metric with multiple features is left with a single feature, it's categorized as "trimmable", because - 
```
is_trimmable = get_effective_feature_count() > 1
```

and so it will be 1, and will be fully removed

**Review by tomtalp — APPROVED** [review]
